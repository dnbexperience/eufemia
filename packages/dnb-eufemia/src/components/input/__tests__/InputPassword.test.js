/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  fakeProps,
  toJson,
  axeComponent,
} from '../../../core/jest/jestSetup'
import Component from '../InputPassword'

import nbNO from '../../../shared/locales/nb-NO'
import enGB from '../../../shared/locales/en-GB'

const nb = nbNO['nb-NO'].Input
const en = enGB['en-GB'].Input

const snapshotProps = {
  ...fakeProps(require.resolve('../InputPassword'), {
    all: true,
    optional: true,
  }),
  input_element: null,
  disabled: false,
}
snapshotProps.id = 'input'
snapshotProps.autocomplete = 'off'

describe('InputPassword component', () => {
  // compare the snapshot
  it('have to match snapshot', () => {
    const Comp = mount(<Component {...snapshotProps} value="test" />)
    expect(toJson(Comp)).toMatchSnapshot()
  })

  // then test the state management
  const Comp = mount(<Component id="input" />)

  it('has correct type by default', () => {
    expect(Comp.find('.dnb-input__input').prop('type')).toBe('password')
  })

  it('has correct state after "focus" trigger', () => {
    Comp.find('input').simulate('focus')
    expect(Comp.find('.dnb-input').prop('data-input-state')).toBe('focus')
  })

  it('has correct aria-label', () => {
    const Comp = mount(<Component id="input" />)

    expect(Comp.find('button').prop('aria-label')).toBe(nb.show_password)

    Comp.setProps({
      lang: 'en-GB',
    })

    expect(Comp.find('button').instance().getAttribute('aria-label')).toBe(
      en.show_password
    )

    expect(Comp.find('button').instance().getAttribute('aria-label')).toBe(
      en.show_password
    )
  })

  it('has aria-describedby and aria-controls', () => {
    Comp.find('input').simulate('focus')
    expect(Comp.find('.dnb-input__input').prop('aria-describedby')).toBe(
      'input-submit-button'
    )
    expect(
      Comp.find('button#input-submit-button').prop('aria-controls')
    ).toBe('input')
  })

  it('has a submit button which gets focus', () => {
    const Comp = mount(<Component />)

    const Button = Comp.find('InputSubmitButton').find('button')
    expect(Button.exists()).toBe(true)

    Button.simulate('focus')
    expect(
      Comp.find('InputSubmitButton')
        .find('.dnb-input__submit-button')
        .prop('data-input-state')
    ).toBe('focus')
  })

  it('can change the visibility of the password', () => {
    const Comp = mount(<Component />)

    const Button = Comp.find('InputSubmitButton').find('button')
    expect(Button.exists()).toBe(true)

    Button.simulate('click')
    expect(Comp.find('.dnb-input__input').prop('type')).toBe('text')

    Button.simulate('click')
    expect(Comp.find('.dnb-input__input').prop('type')).toBe('password')

    expect(
      Comp.find('InputSubmitButton')
        .find('.dnb-input__submit-button')
        .prop('data-input-state')
    ).not.toBe('focus')
  })

  it('events gets triggered on interaction', () => {
    const on_show_password = jest.fn()
    const on_hide_password = jest.fn()
    const Comp = mount(
      <Component
        on_show_password={on_show_password}
        on_hide_password={on_hide_password}
      />
    )

    const Button = Comp.find('InputSubmitButton').find('button')

    Button.simulate('click')
    expect(on_show_password).toBeCalledTimes(1)
    expect(on_hide_password).not.toBeCalled()

    Button.simulate('click')
    expect(on_show_password).toBeCalledTimes(1)
    expect(on_hide_password).toBeCalledTimes(1)

    Button.simulate('click')
    expect(on_show_password).toBeCalledTimes(2)
    expect(on_hide_password).toBeCalledTimes(1)
  })

  it('should validate with ARIA rules as a input with a label', async () => {
    const InputPasswordComp = mount(
      <Component id="input" label="label" value="some value" />
    )
    expect(await axeComponent(InputPasswordComp)).toHaveNoViolations()
  })
})
