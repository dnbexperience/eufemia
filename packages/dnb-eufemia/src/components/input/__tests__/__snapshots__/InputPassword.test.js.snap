// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputPassword component have to match snapshot 1`] = `
<InputPassword
  0={
    Object {
      "displayName": "InputPassword",
      "props": Object {
        "hide_password": "hide_password",
        "on_hide_password": [Function],
        "on_show_password": [Function],
        "show_password": "show_password",
      },
    }
  }
  align={null}
  autocomplete="off"
  class={null}
  className={null}
  clear={null}
  clear_button_title={null}
  custom_element={null}
  custom_method={null}
  disabled={false}
  global_status_id={null}
  hide_password={null}
  icon={null}
  icon_position="left"
  icon_size={null}
  id="input"
  inner_ref={null}
  input_attributes={null}
  input_class={null}
  input_element={null}
  input_state={null}
  keep_placeholder={null}
  label={null}
  label_direction={null}
  label_sr_only={null}
  on_blur={null}
  on_change={null}
  on_focus={null}
  on_hide_password={null}
  on_show_password={null}
  on_state_update={null}
  on_submit={null}
  on_submit_blur={null}
  on_submit_focus={null}
  placeholder={null}
  readOnly={false}
  selectall={null}
  show_password={null}
  size={null}
  skeleton={null}
  status={null}
  status_no_animation={null}
  status_props={null}
  status_state="error"
  stretch={null}
  submit_button_icon="loupe"
  submit_button_status={null}
  submit_button_title={null}
  submit_button_variant="secondary"
  submit_element={null}
  suffix={null}
  type="text"
  value="test"
>
  <Input
    0={
      Object {
        "displayName": "InputPassword",
        "props": Object {
          "hide_password": "hide_password",
          "on_hide_password": [Function],
          "on_show_password": [Function],
          "show_password": "show_password",
        },
      }
    }
    align={null}
    aria-describedby="input-submit-button"
    autocomplete="off"
    class={null}
    className="dnb-input--password"
    clear={null}
    clear_button_title={null}
    custom_element={null}
    custom_method={null}
    disabled={false}
    global_status_id={null}
    hide_password={null}
    icon={null}
    icon_position="left"
    icon_size={null}
    id="input"
    inner_ref={
      Object {
        "current": <input
          aria-describedby="input-submit-button"
          autocomplete="off"
          class="dnb-input__input"
          id="input"
          name="input"
          type="password"
          value="test"
        />,
      }
    }
    input_attributes={null}
    input_class={null}
    input_element={null}
    input_state={null}
    keep_placeholder={null}
    label={null}
    label_direction={null}
    label_sr_only={null}
    on_blur={null}
    on_change={null}
    on_focus={null}
    on_hide_password={null}
    on_show_password={null}
    on_state_update={null}
    on_submit={null}
    on_submit_blur={null}
    on_submit_focus={null}
    placeholder={null}
    readOnly={false}
    selectall={null}
    show_password={null}
    size={null}
    skeleton={null}
    status={null}
    status_no_animation={null}
    status_props={null}
    status_state="error"
    stretch={null}
    submit_button_icon="loupe"
    submit_button_status={null}
    submit_button_title={null}
    submit_button_variant="secondary"
    submit_element={
      <ForwardRef
        aria-controls="input"
        aria-label="Vis passord"
        icon={[Function]}
        id="input-submit-button"
        onClick={[Function]}
        skeleton={null}
        status={null}
        status_state="error"
        type="button"
        variant="secondary"
      />
    }
    suffix={null}
    type="password"
    value="test"
  >
    <span
      className="dnb-input dnb-input--password dnb-form-component dnb-input--password dnb-input--has-submit-element"
      data-has-content="true"
      data-input-state="virgin"
    >
      <span
        className="dnb-input__inner"
      >
        <AlignmentHelper
          className={null}
        >
          <span
            aria-hidden={true}
            className="dnb-alignment-helper"
          />
        </AlignmentHelper>
        <FormStatus
          attributes={null}
          class={null}
          className={null}
          global_status_id={null}
          icon="error"
          icon_size="medium"
          id="input-form-status"
          label={null}
          no_animation={null}
          role={null}
          show={null}
          size="default"
          skeleton={null}
          state="error"
          status="error"
          stretch={null}
          text={null}
          text_id="input-status"
          title={null}
          variant={null}
          width_element={null}
          width_selector={null}
        />
        <span
          className="dnb-input__row"
        >
          <span
            className="dnb-input__shell"
          >
            <input
              aria-describedby="input-submit-button"
              autoComplete="off"
              className="dnb-input__input"
              disabled={false}
              id="input"
              name="input"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              type="password"
              value="test"
            />
          </span>
          <span
            className="dnb-input__submit-element"
          >
            <ForwardRef
              aria-controls="input"
              aria-label="Vis passord"
              icon={[Function]}
              id="input-submit-button"
              onClick={[Function]}
              skeleton={null}
              status={null}
              status_state="error"
              type="button"
              variant="secondary"
            >
              <InputSubmitButton
                aria-controls="input"
                aria-label="Vis passord"
                className={null}
                disabled={false}
                icon={[Function]}
                icon_size={null}
                id="input-submit-button"
                innerRef={null}
                onClick={[Function]}
                on_submit={null}
                on_submit_blur={null}
                on_submit_focus={null}
                skeleton={null}
                status={null}
                status_props={null}
                status_state="error"
                title={null}
                type="button"
                value={null}
                variant="secondary"
              >
                <span
                  className="dnb-input__submit-button"
                  data-input-state="virgin"
                >
                  <Button
                    aria-controls="input"
                    aria-label="Vis passord"
                    bounding={null}
                    class={null}
                    className="dnb-input__submit-button__button dnb-button--input-button"
                    custom_content={null}
                    custom_element={null}
                    custom_method={null}
                    disabled={false}
                    element={null}
                    global_status_id={null}
                    href={null}
                    icon={[Function]}
                    icon_position="right"
                    icon_size={null}
                    id="input-submit-button"
                    innerRef={null}
                    inner_ref={null}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    on_click={null}
                    size={null}
                    skeleton={null}
                    status={null}
                    status_no_animation={null}
                    status_props={null}
                    status_state="error"
                    stretch={null}
                    text={null}
                    title={null}
                    to={null}
                    tooltip={null}
                    type="button"
                    variant="secondary"
                    wrap={null}
                  >
                    <button
                      aria-controls="input"
                      aria-label="Vis passord"
                      className="dnb-button dnb-button--secondary dnb-input__submit-button__button dnb-button--input-button dnb-button--icon-position-right dnb-button--has-icon dnb-button--size-medium"
                      disabled={false}
                      id="input-submit-button"
                      onBlur={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      type="button"
                    >
                      <Content
                        aria-controls="input"
                        aria-label="Vis passord"
                        bounding={null}
                        class={null}
                        className="dnb-input__submit-button__button dnb-button--input-button"
                        content={null}
                        custom_content={null}
                        custom_element={null}
                        custom_method={null}
                        disabled={false}
                        element={null}
                        global_status_id={null}
                        href={null}
                        icon={[Function]}
                        icon_position="right"
                        icon_size={null}
                        id="input-submit-button"
                        innerRef={null}
                        inner_ref={null}
                        isIconOnly={true}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        on_click={null}
                        size={null}
                        skeleton={false}
                        status={null}
                        status_no_animation={null}
                        status_props={null}
                        status_state="error"
                        stretch={null}
                        text={null}
                        title={null}
                        to={null}
                        tooltip={null}
                        type="button"
                        variant="secondary"
                        wrap={null}
                      >
                        <span
                          className="dnb-button__alignment"
                          key="button-text-empty"
                        >
                          ‌
                        </span>
                        <IconPrimary
                          alt={null}
                          aria-hidden={null}
                          attributes={null}
                          border={null}
                          className="dnb-button__icon"
                          color={null}
                          height={null}
                          icon={[Function]}
                          inherit_color={true}
                          key="button-icon"
                          modifier={null}
                          size={null}
                          skeleton={false}
                          title={null}
                          width={null}
                        >
                          <span
                            aria-hidden={true}
                            className="dnb-icon dnb-icon--default dnb-button__icon dnb-icon--inherit-color"
                            data-test-id="view icon"
                            role="presentation"
                          >
                            <view>
                              <svg
                                fill="none"
                                height={16}
                                viewBox="0 0 16 16"
                                width={16}
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M8 4c-2.51-.042-5.103 1.711-6.734 3.506a1.035 1.035 0 0 0 0 1.385c1.595 1.758 4.182 3.55 6.733 3.508 2.551.043 5.138-1.75 6.735-3.508a1.035 1.035 0 0 0 0-1.385C13.101 5.711 10.507 3.958 8 4.001Z"
                                  stroke="#000"
                                  strokeLinecap="round"
                                  strokeLinejoin="round"
                                  strokeWidth={1.5}
                                />
                                <path
                                  d="M9.75 8.2a1.75 1.75 0 1 1-3.5 0 1.75 1.75 0 0 1 3.5 0Z"
                                  stroke="#000"
                                  strokeLinecap="round"
                                  strokeLinejoin="round"
                                  strokeWidth={1.5}
                                />
                              </svg>
                            </view>
                          </span>
                        </IconPrimary>
                      </Content>
                    </button>
                    <FormStatus
                      attributes={null}
                      class={null}
                      className={null}
                      global_status_id={null}
                      icon="error"
                      icon_size="medium"
                      id="input-submit-button-form-status"
                      label={null}
                      no_animation={null}
                      role={null}
                      show={null}
                      size="default"
                      skeleton={null}
                      state="error"
                      status="error"
                      stretch={null}
                      text={null}
                      text_id="input-submit-button-status"
                      title={null}
                      variant={null}
                      width_element={null}
                      width_selector={null}
                    />
                  </Button>
                </span>
              </InputSubmitButton>
            </ForwardRef>
          </span>
        </span>
      </span>
    </span>
  </Input>
</InputPassword>
`;
